observability/micrometer/src/main/java/org/wildfly/extension/micrometer/MicrometerSubsystemAdd.java

```diff
@@ -57,40 +54,37 @@ public class MicrometerSubsystemAdd extends AbstractBoottimeAddStepHandler {
             throws OperationFailedException {
-        // If the MP Metrics module is not installed, we need to install the WF Metrics DPU and initiate a metrics
-        // collection. If MP Metrics *is* installed, then we do not need to do either of those things, as that module
-        // handles that instead.
-        if (!context.getCapabilityServiceSupport().hasCapability(METRICS_SCAN_CAPABILITY)) {
-            List<String> exposedSubsystems = MicrometerSubsystemDefinition.EXPOSED_SUBSYSTEMS.unwrap(context, model);
-            boolean exposeAnySubsystem = exposedSubsystems.remove("*");
-            boolean securityEnabled = MicrometerSubsystemDefinition.SECURITY_ENABLED.resolveModelAttribute(context, model)
-                    .asBoolean();
 
-            MicrometerRegistryService.install(context, securityEnabled);
-            MicrometerCollectorService.install(context);
-            MicrometerContextService.install(context, securityEnabled);
+        List<String> exposedSubsystems = MicrometerSubsystemDefinition.EXPOSED_SUBSYSTEMS.unwrap(context, model);
+        boolean exposeAnySubsystem = exposedSubsystems.remove("*");
+        boolean securityEnabled = MicrometerSubsystemDefinition.SECURITY_ENABLED.resolveModelAttribute(context, model)
+                .asBoolean();
+
+        MicrometerRegistryService.install(context, securityEnabled);
+        MicrometerCollectorService.install(context);
+        MicrometerContextService.install(context, securityEnabled);
+
+        context.addStep(new AbstractDeploymentChainStep() {
+            @Override
+            public void execute(DeploymentProcessorTarget processorTarget) {
+                processorTarget.addDeploymentProcessor(SUBSYSTEM_NAME, DEPENDENCIES, 0x1910,
+                        new MicrometerDependencyProcessor());
+                processorTarget.addDeploymentProcessor(SUBSYSTEM_NAME, POST_MODULE, POST_MODULE_METRICS - 1, // ???
+                        new MicrometerSubsystemDeploymentProcessor(exposeAnySubsystem, exposedSubsystems));
+            }
+        }, RUNTIME);
 
-            context.addStep(new AbstractDeploymentChainStep() {
-                @Override
-                public void execute(DeploymentProcessorTarget processorTarget) {
-                    processorTarget.addDeploymentProcessor(SUBSYSTEM_NAME, DEPENDENCIES, 0x1910,
-                            new MicrometerDependencyProcessor());
-                    processorTarget.addDeploymentProcessor(SUBSYSTEM_NAME, POST_MODULE, POST_MODULE_METRICS - 1, // ???
-                            new MicrometerSubsystemDeploymentProcessor(exposeAnySubsystem, exposedSubsystems));
-                }
-            }, RUNTIME);
+        context.addStep((operationContext, modelNode) -> {
+            ServiceController<?> serviceController = context.getServiceRegistry(false).getService(MICROMETER_COLLECTOR);
+            MicrometerCollector micrometerCollector = MicrometerCollector.class.cast(serviceController.getValue());
```

Comment:
Please file a JIRA to get rid of using the MSC ServiceRegistry this way. MicrometerCollectorService.install can return a Supplier< MicrometerCollector> that gets populated in MicrometerCollectorService.start and cleared in stop.

Reply:
Filed: https://issues.redhat.com/browse/WFLY-17166